linux/devdoc/socket_transport_linux.md

```diff
@@ -0,0 +1,284 @@
+﻿# socket transport Linux
+
+## Overview
+
+socket_transport_linux is the module that abstracts the socket transport for the linux platform
+
+The `SOCKET_TYPE` determines which API are applicable for this instance of socket_transport.  The following chart shows which API's are allowed under which type:
+
+| API                                    | Allowable Socket Type
+|----------------------------------------|----------------------
+| socket_transport_create                | BOTH
+| socket_transport_destroy               | BOTH
+| socket_transport_connect               | SOCKET_CLIENT
+| socket_transport_listen                | SOCKET_SERVER
+| socket_transport_disconnect            | BOTH
+| socket_transport_accept                | SOCKET_SERVER
+| socket_transport_send                  | BOTH
+| socket_transport_receive               | BOTH
+| socket_transport_get_underlying_socket | BOTH
+
+## Exposed API
+
+```c
+typedef struct SOCKET_TRANSPORT_TAG* SOCKET_TRANSPORT_HANDLE;
+
+#define SOCKET_SEND_RESULT_VALUES \
+    SOCKET_SEND_OK, \
+    SOCKET_SEND_ERROR, \
+    SOCKET_SEND_SHUTDOWN
+
+MU_DEFINE_ENUM(SOCKET_SEND_RESULT, SOCKET_SEND_RESULT_VALUES)
+
+#define SOCKET_RECEIVE_RESULT_VALUES \
+    SOCKET_RECEIVE_OK, \
+    SOCKET_RECEIVE_WOULD_BLOCK, \
+    SOCKET_RECEIVE_ERROR, \
+    SOCKET_RECEIVE_SHUTDOWN
+
+MU_DEFINE_ENUM(SOCKET_RECEIVE_RESULT, SOCKET_RECEIVE_RESULT_VALUES)
+
+#define SOCKET_TYPE_VALUES \
+    SOCKET_CLIENT, \
+    SOCKET_SERVER
+
+MU_DEFINE_ENUM(SOCKET_TYPE, SOCKET_TYPE_VALUES)
+
+typedef struct SOCKET_BUFFER_TAG
+{
+    uint32_t length;
+    void* buffer;
```

Comment:
>void [](http://example.com/codeflow?start=4&length=4)

let's have this as "unsigned char*'

Also since this "sort of" moves data, each SOCKET_BUFFER owns its data? Brings up the question of how to avoid (if possible) memory copies (if any...)

Reply:
We shall try to not copy anything.